src/topology/inseparable.lean

```diff
@@ -178,12 +179,197 @@ def specialization_preorder : preorder X :=
+  (H : ∀ i, stable_under_specialization (S i)) : stable_under_specialization (⋂ i, S i) :=
+is_lower_set_Inter H
+
+lemma stable_under_generalization_Union {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_generalization (S i)) : stable_under_generalization (⋃ i, S i) :=
+is_upper_set_Union H
+
+lemma stable_under_generalization_Inter {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_generalization (S i)) : stable_under_generalization (⋂ i, S i) :=
+is_upper_set_Inter H
+
+lemma stable_under_specialization_iff_eq_Union {s : set X} :
+  stable_under_specialization s ↔ s = ⋃ x ∈ s, closure {x} :=
+show is_lower_set s ↔ _, by simp only [closure_singleton_eq_Iic, ← lower_closure_eq,
+  ← lower_set.supr_Iic, lower_set.coe_supr₂, lower_set.coe_Iic, eq_comm]
+
+alias stable_under_specialization_iff_eq_Union ↔ stable_under_specialization.eq_Union _
+
+/-- A set is stable under specialization iff it is a union of closed sets. -/
+lemma stable_under_specialization_iff_exists_sUnion_eq {s : set X} :
+  stable_under_specialization s ↔ ∃ (S : set (set X)), (∀ s ∈ S, is_closed s) ∧ ⋃₀ S = s :=
+begin
+  refine ⟨λ H, ⟨(λ x : X, closure {x}) '' s, _, _⟩, λ ⟨S, hS, e⟩, e ▸
+    stable_under_specialization_sUnion S (λ x hx, (hS x hx).stable_under_specialization)⟩,
+  { rintros _ ⟨_, _, rfl⟩, exact is_closed_closure },
+  { conv_rhs { rw stable_under_specialization_iff_eq_Union.mp H }, simp }
+end
+
+/-- A set is stable under generalization iff it is an intersection of open sets. -/
+lemma stable_under_generalization_iff_exists_sInter_eq {s : set X} :
+  stable_under_generalization s ↔ ∃ (S : set (set X)), (∀ s ∈ S, is_open s) ∧ ⋂₀ S = s :=
+begin
+  refine ⟨_, λ ⟨S, hS, e⟩, e ▸
+    stable_under_generalization_sInter S (λ x hx, (hS x hx).stable_under_generalization)⟩,
+  rw [← stable_under_specialization_compl_iff, stable_under_specialization_iff_exists_sUnion_eq],
+  exact λ ⟨S, h₁, h₂⟩, ⟨has_compl.compl '' S, λ s ⟨t, ht, e⟩, e ▸ (h₁ t ht).is_open_compl,
+    compl_injective ((sUnion_eq_compl_sInter_compl S).symm.trans h₂)⟩
+end
+
+lemma stable_under_specialization.preimage {s : set Y}
+  (hs : stable_under_specialization s) (hf : continuous f) :
+  stable_under_specialization (f ⁻¹' s) :=
+is_lower_set.preimage hs hf.specialization_monotone
+
+lemma stable_under_generalization.preimage {s : set Y}
+  (hs : stable_under_generalization s) (hf : continuous f) :
+  stable_under_generalization (f ⁻¹' s) :=
+is_upper_set.preimage hs hf.specialization_monotone
+
+/-- A map `f` between topological spaces is specializing if specializations lifts along `f`,
+i.e. for each `f x' ⤳ y` there is some `x` with `x' ⤳ x` whose image is `y`. -/
+def specializing (f : X → Y) : Prop :=
+relation.fibration (flip (⤳)) (flip (⤳)) f
+
+/-- A map `f` between topological spaces is generalizing if generalizations lifts along `f`,
+i.e. for each `y ⤳ f x'` there is some `x ⤳ x'` whose image is `y`. -/
+def _root_.generalizing (f : X → Y) : Prop :=
```

Comment:
This sounds like a pretty bad name to drop in the root namespace. What about namespace `topology` (or `topological_space`, but that's longer)?

Reply:
But other predicates on maps between topological spaces are in the root namespace as well. Would renaming it into `generalizing_map` make sense?

Reply:
`embedding` is in fact a pretty bad name to have too. This is an historical accident and the solution would be to move all of them into a namespace.

Reply:
Since I think this change will be more controversial, I'll ask on Zulip before making the PR. In the mean time, I have appended `_map` to both `generalizing` and `specializing`.